src/geode/mesh/helpers/gradient_computation.cpp

```diff
@@ -0,0 +1,137 @@
+/*
+ * Copyright (c) 2019 - 2024 Geode-solutions
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ */
+
+#include <geode/mesh/helpers/regular_grid_point_function.h>
+
+#include <geode/basic/attribute_manager.h>
+#include <geode/basic/logger.h>
+#include <geode/basic/pimpl_impl.h>
+
+#include <geode/geometry/distance.h>
+#include <geode/geometry/point.h>
+#include <geode/geometry/vector.h>
+
+#include <geode/mesh/core/solid_mesh.h>
+#include <geode/mesh/core/surface_mesh.h>
+#include <geode/mesh/helpers/private/regular_grid_shape_function.h>
+
+namespace
+{
+    template < typename Mesh >
+    void compute_scalar_function_gradient( const Mesh& mesh,
+        absl::string_view scalar_function_name,
+        absl::string_view output_gradient_attribute_name )
+    {
+        OPENGEODE_EXCEPTION( mesh.vertex_attribute_manager().attribute_exists(
+                                 scalar_function_name ),
+            "[compute_scalar_function_gradient] No attribute exists with given "
+            "name." );
+        OPENGEODE_EXCEPTION( mesh.vertex_attribute_manager().attribute_type(
+                                 scalar_function_name )
+                                 == typeid( double ).name(),
+            "[compute_scalar_function_gradient] The attribute linked to given "
+            "name is not scalar." );
+        OPENGEODE_EXCEPTION( !mesh.vertex_attribute_manager().attribute_exists(
```

Comment:
If the attribute should not exist, does the function name should be create_XXX or compute_XXX?